### DISCUSSION TITLE
command not found: brew

### DISCUSSION THREAD

**Discussion by sav5000** [tagatac/bagoup#49] [Q&A] [answered]
Hi, I love the idea of your program and it's exactly what I've been looking for. I'm completely new to using coding and using Terminal, though, and am having trouble getting started with this. I downloaded the code and also wkhtmltopdf since I want to export all my texts with photos to PDF. I opened Terminal and put in both "brew install wkhtmltopdf" and "brew tap tagatac/bagoup
brew install bagoup" but it is saying "command not found" in response to both of them. I'm on MacOS Monterey 12.6.7. I have no idea what else to do — would really appreciate any help!! Thanks so much!

**Comment by tagatac** [comment, marked as answer]
You likely don't have homebrew installed. homebrew is the standard for package management on Mac OS these days. You can read more about it, including installation instructions here: https://brew.sh/

In case that doesn't resolve your issue, feel free to post back with the full error message you get when you try to run
```
brew install wkhtmltopdf
```

**Comment by sav5000** [comment]
Thank you so much, I installed it and was also able to install your program! I would never have figured that out — I really appreciate your help! I'm trying to export my messages now but keep getting an error that the vcard file from a contact (any contact) doesn't exist. At first it was because I didn't know how to put a space in between the name correctly (told you I'm a newbie haha) so I looked it up and added a backslash, but it still is not finding the card. Does this look right to you?
`bagoup --contacts-path John\ Smith.vcf --export-path messages-export-pdf --pdf  --include-ppa --copy-attachments
`

I get the error:
`ERROR: get contacts from vcard file "John Smith.vcf": open John Smith.vcf: no such file or directory`

All the people I text are in my contacts, so I'm not sure what else to do or why it's not finding the contact cards. Thank you again!

**Reply by tagatac** [reply]
Awesome. It sounds like we're close. From the error message, it seems like you are not passing the correct path to the vcard file. It needs to be the relative path from where you're running the bagoup command, or the full path, e.g. `/Users/sav5000/Desktop/John\ Smith.vcf`.

BTW, it is recommended to use a vcf file with all of your contacts. A vcf file for a single contact will only label conversations with that contact, and not the others. Both Google Contacts and Mac OS Contacts are able to export all contacts into a single file.
- https://support.google.com/contacts/answer/7199294?hl=en&co=GENIE.Platform%3DDesktop
- https://support.apple.com/guide/contacts/export-or-archive-contacts-adrbdcfd32e6/mac#:~:text=Export%20all%20contacts,abbu)%20only.

**Reply by sav5000** [reply]
Alright, I added the full contact path to the command — this info helped a ton and I've made lots of progress and am excited!! I've spent months looking for a way to back up my messages, so to finally get somewhere is a relief! I am still getting some errors and issues though which I've described below. I am so appreciative of your help and patience — and I understand if this is too much lol, just let me know!

After exporting the vcf to the desktop, I put in the name of one specific person in the Terminal command, and it did export that one, but didn't attach the name. It also exported a handful of other random phone numbers at the same time that I don't recognize lol, and have no messages with. I exported my most used contacts into one vcf file on the desktop, specified that in the Terminal command, and the program exported two of them (plus the other random ones), but still not with their names — and one of them doesn't include messages after 2022 for some reason.

So how would I export messages from just those selected people in my exported vcf file? Additionally, how would I export a group chat?

I am frequently getting these types of error messages as well. The 555 numbers are placeholders for actual contacts' numbers.

`<WARN: no valid text or attributedBody for message 32163`

`WARN: chat file "messages-export-pdf/+15555555555/iMessage;-;+155555555555;;;SMS;-;+15555555555.pdf" - message 26855 - image/jpeg attachment "Screen Shot 2021-01-31 at 10.04.50 AM.jpeg" (ID 3179) - attachment has no local filename`

Thank you so much again!

**Reply by tagatac** [reply]
Sounds like some good progress 🙂

> So how would I export messages from just those selected people in my exported vcf file?

This is not a feature of bagoup at the moment. The tool will always export all of the conversations in your chat database. The contacts specified in your vCard file only determine which of the conversations will be labeled (if a match is found). Of course it is possible to delete folders for contacts you are not interested in exporting.

> Additionally, how would I export a group chat?

Group chats are always included in the export. Named group chats will be in folders with the same name. Unnamed group chats will be in folders named with the chat ID from the database file (usually the concatenation of "chat" and a number, e.g. `chat12345678901234567`).

> `WARN: no valid text or attributedBody for message 32163`

This is just a warning, and there's nothing really for you to do. I don't have a full list of the reasons for blank messages, but at least sharing a location seems to be one reason. I've opened #50 to add more context to the export instead of just printing a warning.

> `WARN: chat file "messages-export-pdf/+15555555555/iMessage;-;+155555555555;;;SMS;-;+15555555555.pdf" - message 26855 - image/jpeg attachment "Screen Shot 2021-01-31 at 10.04.50 AM.jpeg" (ID 3179) - attachment has no local filename`

This warning arises when the chat database knows about an attachment, but does not have a location for it on your hard drive. You can try opening the conversation in Messages and scrolling to the attachment to encourage Messages to download the attachment from the cloud. But if it's been a while, it's likely that you will not be able to retrieve the file.

> After exporting the vcf to the desktop, I put in the name of one specific person in the Terminal command, and it did export that one, but didn't attach the name.

The matching looks at phone numbers (with [parentheses, dashes, and spaces removed](https://github.com/tagatac/bagoup/blob/8d4c9951241d344bf911e57988337b64fea0d506/opsys/opsys.go#L169)) and email addresses. If you are not getting a match you expect, take a look at the contact card and make sure the phone number or email address matches the one used to label the exported conversation.

> It also exported a handful of other random phone numbers at the same time that I don't recognize lol, and have no messages with.

It is unlikely that bagoup would export numbers that you haven't messaged, as all of the information is pulled from the chat database on your computer. Have you looked at the export files to see the contents of these conversations? Perhaps these are unnamed group chats that I described above?

> I exported my most used contacts into one vcf file on the desktop, specified that in the Terminal command, and the program exported two of them (plus the other random ones), but still not with their names — and one of them doesn't include messages after 2022 for some reason.

Again, take a look at the contact cards to make sure that the numbers match. Notably, if the contact card includes a country code and the chat database does not (or vice versa), a match will not be made.

WRT the missing messages, this does sound like a possible bug, but it is difficult to debug without the `chat.db` file. A couple of questions:
1. Did you get a results printout at the end of the run? If yes, can you post it here? It will look something like this:
```
BAGOUP RESULTS:
bagoup version: 2.3 Darwin/arm64
Export folder: "messages-export"
Export files written: 121
Chats exported: 124
Valid messages exported: 14442
Invalid messages exported (see warnings above): 15
Attachments copied: 0
Attachments referenced or embedded: 854
	image/png: 24
	image/heic: 202
	text/x-vlocation: 8
	video/quicktime: 26
	text/vcard: 3
	application/octet-stream: 426
	image/tiff: 4
	application/pdf: 2
	image/jpeg: 129
	image/gif: 30
Attachments embedded: 0
Attachments missing (see warnings above): 3
HEIC conversions completed: 0
HEIC conversions failed (see warnings above): 0
Time elapsed: 2m0.869446042s
```
2. Can you confirm that you have messages with this contact after 2022 in the Messages app?

The easiest way to debug this is for you to send me your `chat.db` file at david@tagatac.net, but I understand it's a sensitive file, and I would never do that myself. If you can convince yourself that bagoup is dropping some messages that it should be exporting, it would be great if you could open a bug with as much relevant detail as possible. I may ask you to run a debug version of bagoup to get more information about the bug while minimizing how much sensitive data you need to share.

**Reply by sav5000** [reply]
Ok, I think it worked this time! :) I was able to export all the chats. A few things I noticed:
I had deleted all iMessages from before 2022, but they were exported anyway, some going back to 2013 for some reason — including lots of empty folders from spam numbers I had also already deleted. I know when you delete things on a Mac, they're usually only deleted on the surface level and the files are still there somewhere. So if I were to export more messages in the future, I guess the program would still find and export all messages even if I deleted them in iMessage, unless I deleted the chat.db file, correct?

The chat that was missing messages added more in another file the longer the program ran. In the first incomplete file, the contact name still wasn't there even though the phone number matched the contact card. In the second file (which completed the messages we sent each other up to yesterday), the name did show up because it was associated with an email instead. Don't know why it separated them since all the info is in the same contact card, but I'm not too concerned since I'm guessing it's an Apple issue. It's just great to finally be able to export these messages!

I did also find the group chat. Again, one of the contact's names showed up and the other didn't, but it's not a big deal. I exported all the data I needed which is just such a big relief. Thank you so much again — I so appreciate your addressing everything and your time, help, and patience.

Here is the results printout if you still need it (sorry it's not in code format, I tried but it lumped it all together in one paragraph/block of text lol)
2023/07/10 08:43:51 BAGOUP RESULTS:

bagoup version: 2.3 Darwin/x86_64
Export folder: "messages-export5-pdf"
Export files written: 67
Chats exported: 76
Valid messages exported: 18377
Invalid messages exported (see warnings above): 154
Attachments copied: 0
Attachments referenced or embedded: 3300
	video/quicktime: 51
	application/epub+zip: 2
	image/jpeg: 1145
	image/png: 92
	image/tiff: 3
	application/pdf: 5
	audio/x-m4a: 1
	image/webp: 4
	application/octet-stream: 1878
	image/gif: 98
	audio/amr: 1
	application/paprikarecipes: 20
Attachments embedded: 2998
	application/octet-stream: 1855
	image/png: 92
	image/jpeg: 957
	image/tiff: 3
	image/gif: 87
	image/webp: 4
Attachments missing (see warnings above): 212
HEIC conversions completed: 394
HEIC conversions failed (see warnings above): 0
Time elapsed: 35m4.578706449s

**Reply by tagatac** [reply]
Glad to hear you got some utility out of bagoup! And thanks so much for your donation 🙏

> So if I were to export more messages in the future, I guess the program would still find and export all messages even if I deleted them in iMessage, unless I deleted the chat.db file, correct?

I have not experimented too much with deleting messages, but it does seem like deleting messages in Mac OS 13 does delete the messages from `chat.db`. But Apple does change the mechanics of the Messages app from time to time, so it may depend on what version of Mac OS you were running when you deleted the messages. Also, my anecdotal observations may not be universally accurate.

> The chat that was missing messages added more in another file the longer the program ran. In the first incomplete file, the contact name still wasn't there even though the phone number matched the contact card. In the second file (which completed the messages we sent each other up to yesterday), the name did show up because it was associated with an email instead.

This is a little odd, but I wouldn't read too much into it, as weird things can happen on incomplete runs.

> I did also find the group chat. Again, one of the contact's names showed up and the other didn't

And the contact card matches the phone number or email address in the export? This sounds like it could be a bug, and I'd love to dig into it. If you could create a new bug with as much detail as possible, that would be great.

**Reply by sav5000** [reply]
The contact card does match both the phone number and email in the export. I can give you as much info as I have about this, but it’s not much — and I think it’s an iCloud/Apple issue, not a bug with bagoup, so I don’t think it’s worth opening a new bug for. But let me know if you want me to copy/paste what I’ve described here into a new one anyway!

I don’t really understand how Apple does things sometimes, but I think it’s regarding selecting either your phone number or email to start new conversations from in iMessage settings. The contact card for the person I was messaging (the one that was split up into two files, and whose name didn’t attach to the phone number in the group chat) included his phone number and email associated with iCloud. Originally he was texting me through his phone number, which is associated with the first message file I exported. Then he said he had to reset his Apple password and Apple automatically switched some things for him, so messages are now sending from his Apple ID/email. Still not sure how this works, because he was texting me with his phone number as well, like when messages had to be sent as SMS instead of iMessage. But the second message thread export with him were all under his Apple ID email, thus matching up his name. 

So since both his phone number and email were included in the contact card, bagoup seems like it should have grouped them together. But the only messages it exported with a name associated with it were because the messages were sent from an email instead of phone number, I guess. 

One other thing worth mentioning is that I have multiple emails and phone numbers included in all my contact cards, though I only text their cell (instead of their home phone number, for example) and only one of their emails is associated with their Apple ID. Maybe this affects the program associating names/numbers, because it has too many to choose from, even though only one is being used for messaging? This is all I know — sorry if it’s confusing, feel free to disregard it, and let me know if you need other info!

**Comment by sav5000** [comment]
Hey again, you helped me last year when I was figuring out your program. Thanks again! The time has come again to export my messages, and I am running into some different problems this time. Still know nothing about coding, so bear with me, but I haven’t been able to figure out how to fix these couple issues myself after spending quite a while on it. I’d really appreciate any further help if you can!

I followed the same steps as I did before but am running into two main issues. Firstly, the program doesn’t seem to have exported all the conversations I’ve had, specifically group texts and messages from specific people I’ve texted plenty of times. I reran the command a couple times and then another contact’s convo thread finally showed up, but there are still some other convos missing. Don’t know why this would be? 

The other thing is that all the convos exported perfectly except for one, which shows squares with question marks instead of text/words. Emojis show up fine, but all the text is question marks (see attached pic).

The program in Terminal ends with “Done” and then “Exit with code 1 due to network error: ContentAccessDenied.”

Again, I really appreciate your help and program!
<img width="453" alt="Screen Shot 2024-09-05 at 6 38 43 PM" src="https://github.com/user-attachments/assets/8c83496a-e6cd-4b1b-8a71-448abc1cd2c9">

**Reply by tagatac** [reply]
Hi, lets figure out what that error is about. Can you paste the entire error message ending with `Exit with code 1 due to network error: ContentAccessDenied.`?

**Reply by sav5000** [reply]
That was the only error like that, which seems to relate to the wkhtmiltopdf program. There are many other errors all having to do with attachments not being found, which is fine. One of the main repeating errors is Warning: Failed to load file:///var/folders/mp/h_y9z0393p3c0_46mq0hz5_m0000gn/T/bagoup847734668/IMG_1181.jpeg (ignore). The entire error message is the whole bagoup process and is very long, but I can still post it if it would help. It mostly just consists of the attachment errors.

**Reply by tagatac** [reply]
I've opened an issue for each of these. Let's continue the conversation for each separately on those open issues:
- #62 
- #63
